server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentTracker.java

```diff
@@ -0,0 +1,425 @@
+        failures.record(true);
+    }
+
+    long getTotalUploadsSucceeded() {
+        return totalUploadsSucceeded;
+    }
+
+    void incrementTotalUploadSucceeded() {
+        totalUploadsSucceeded += 1;
+        failures.record(false);
+    }
+
+    long getInflightUploads() {
+        return totalUploadsStarted - totalUploadsFailed - totalUploadsSucceeded;
+    }
+
+    long getRejectionCount() {
+        return rejectionCount.get();
+    }
+
+    void incrementRejectionCount() {
+        rejectionCount.incrementAndGet();
+    }
+
+    void incrementRejectionCount(String rejectionReason) {
+        rejectionCountMap.computeIfAbsent(rejectionReason, k -> new AtomicLong()).incrementAndGet();
+    }
+
+    long getRejectionCount(String rejectionReason) {
+        return rejectionCountMap.get(rejectionReason).get();
+    }
+
+    Map<String, Long> getLatestLocalFileNameLengthMap() {
+        return latestLocalFileNameLengthMap;
+    }
+
+    void setLatestLocalFileNameLengthMap(Map<String, Long> latestLocalFileNameLengthMap) {
+        this.latestLocalFileNameLengthMap = latestLocalFileNameLengthMap;
+        computeBytesLag();
+    }
+
+    void addToLatestUploadFiles(String file) {
+        this.latestUploadFiles.add(file);
+        computeBytesLag();
+    }
+
+    private void computeBytesLag() {
+        if (latestLocalFileNameLengthMap == null || latestLocalFileNameLengthMap.isEmpty()) {
+            return;
+        }
+        Set<String> filesNotYetUploaded = latestLocalFileNameLengthMap.keySet()
+            .stream()
+            .filter(f -> !latestUploadFiles.contains(f))
+            .collect(Collectors.toSet());
+        this.bytesLag = filesNotYetUploaded.stream().map(latestLocalFileNameLengthMap::get).mapToLong(Long::longValue).sum();
+    }
+
+    int getConsecutiveFailureCount() {
+        return failures.length();
+    }
+
+    boolean isUploadBytesAverageReady() {
+        return uploadBytesMovingAverageReference.get().isReady();
+    }
+
+    double getUploadBytesAverage() {
+        return uploadBytesMovingAverageReference.get().getAverage();
+    }
+
+    void addUploadBytes(long size) {
+        synchronized (uploadBytesMutex) {
+            this.uploadBytesMovingAverageReference.get().record(size);
+        }
+    }
+
+    /**
+     * Updates the window size for data collection of upload bytes. This also resets any data collected so far.
+     *
+     * @param updatedSize the updated size
+     */
+    void updateUploadBytesMovingAverageWindowSize(int updatedSize) {
+        synchronized (uploadBytesMutex) {
+            this.uploadBytesMovingAverageReference.set(this.uploadBytesMovingAverageReference.get().copyWithSize(updatedSize));
+        }
+    }
+
+    boolean isUploadBytesPerSecAverageReady() {
+        return uploadBytesPerSecMovingAverageReference.get().isReady();
+    }
+
+    double getUploadBytesPerSecAverage() {
+        return uploadBytesPerSecMovingAverageReference.get().getAverage();
```

Comment:
The underlying moving average isn't thread safe, so I'm not sure that this is guaranteed to reflect the latest values unless it is protected by the same synchronized blocks as the methods that update it.

Reply:
AtomicReference are inherently thread safe. The update of the underlying referenced object and changing the reference is controlled through synchronisation for thread-safe updates.

Reply:
The _reference_ is thread safe (a volatile reference would give you the same guarantees). This call is guaranteed to see the correct reference, and the external synchronization ensures that writes are applied atomically to the MovingAverage instance, but I think this call could happen after an update and still see an older value of the underlying `average` double within the MovingAverage instance because it is not volatile and there is no other synchronization here.

Reply:
> I think this call could happen after an update and still see an older value of the underlying average double within the MovingAverage instance because it is not volatile and there is no other synchronization here.

Lets say an update happened which means the AtomicReference object's reference has been updated atomically. Now if we try to do a get(), it should always give the correct instance. Sharing the java doc of the get method.
 
```
   /**
     * Returns the current value,
     * with memory effects as specified by {@link VarHandle#getVolatile}.
     *
     * @return the current value
     */
    public final V get() {
        return value;
    }
```

Reply:
I agree you will always see the correct reference (I also think you'd get the same guarantees from a plain volatile reference). However, this is the problem I'm referring to:

Thread 1: tracker.addUploadBytes()->movingAverage.record()->writes new value to the movingAverage.average field
Thread 2: tracker.getUploadBytesAverage()->movingAverage.getAverage()-> reads value from movingAverage.average field

If my understanding of the JDK's memory model is correct, there's no memory barrier for thread 2's read of movingAverage.average, so there's no guarantee it will see the result written by thread 1. I believe making the `average` field volatile would solve that problem.

Reply:
Got it. That's true. I thought we are talking about the atomic reference object. Will make the change.